File: internal/naginata/table.go
package naginata

// Entry は同時押しの組み合わせと出力文字の対応を表す。
type Entry struct {
	Keys KeySet
	Text string
}

// 制御用の出力文字。
const (
	TextBackspace = "\b"
	TextEnter     = "\n"
)

// Table は薙刀式v15のかな変換テーブル。
// eswai/qmk_firmware users/naginata_v15/naginata_v15.c の ngmap を移植した。
// 編集モードとエンター付き句読点の入力は練習対象外なので含めない。
var Table = []Entry{
	// 清音
	{Set(KeyJ), "あ"},
	{Set(KeyK), "い"},
	{Set(KeyL), "う"},
	{Set(KeySpace, KeyO), "え"},
	{Set(KeySpace, KeyN), "お"},
	{Set(KeyF), "か"},
	{Set(KeyW), "き"},
	{Set(KeyH), "く"},
	{Set(KeyS), "け"},
	{Set(KeyV), "こ"},
	{Set(KeySpace, KeyU), "さ"},
	{Set(KeyR), "し"},
	{Set(KeyO), "す"},
	{Set(KeySpace, KeyA), "せ"},
	{Set(KeyB), "そ"},
	{Set(KeyN), "た"},
	{Set(KeySpace, KeyG), "ち"},
	{Set(KeySpace, KeyL), "つ"},
	{Set(KeyE), "て"},
	{Set(KeyD), "と"},
	{Set(KeyM), "な"},
	{Set(KeySpace, KeyD), "に"},
	{Set(KeySpace, KeyW), "ぬ"},
	{Set(KeySpace, KeyR), "ね"},
	{Set(KeySpace, KeyJ), "の"},
	{Set(KeyC), "は"},
	{Set(KeyX), "ひ"},
	{Set(KeySpace, KeyX), "ひ"},
	{Set(KeySpace, KeySemi), "ふ"},
	{Set(KeyP), "へ"},
	{Set(KeyZ), "ほ"},
	{Set(KeySpace, KeyZ), "ほ"},
	{Set(KeySpace, KeyF), "ま"},
	{Set(KeySpace, KeyB), "み"},
	{Set(KeySpace, KeyComma), "む"},
	{Set(KeySpace, KeyS), "め"},
	{Set(KeySpace, KeyK), "も"},
	{Set(KeySpace, KeyH), "や"},
	{Set(KeySpace, KeyP), "ゆ"},
	{Set(KeySpace, KeyI), "よ"},
	{Set(KeyDot), "ら"},
	{Set(KeySpace, KeyE), "り"},
	{Set(KeyI), "る"},
	{Set(KeySlash), "れ"},
	{Set(KeySpace, KeySlash), "れ"},
	{Set(KeyA), "ろ"},
	{Set(KeySpace, KeyDot), "わ"},
	{Set(KeySpace, KeyC), "を"},
	{Set(KeyComma), "ん"},
	{Set(KeySemi), "ー"},

	// 濁音
	{Set(KeyJ, KeyF), "が"},
	{Set(KeyJ, KeyW), "ぎ"},
	{Set(KeyF, KeyH), "ぐ"},
	{Set(KeyJ, KeyS), "げ"},
	{Set(KeyJ, KeyV), "ご"},
	{Set(KeyF, KeyU), "ざ"},
	{Set(KeyJ, KeyR), "じ"},
	{Set(KeyF, KeyO), "ず"},
	{Set(KeyJ, KeyA), "ぜ"},
	{Set(KeyJ, KeyB), "ぞ"},
	{Set(KeyF, KeyN), "だ"},
	{Set(KeyJ, KeyG), "ぢ"},
	{Set(KeyF, KeyL), "づ"},
	{Set(KeyJ, KeyE), "で"},
	{Set(KeyJ, KeyD), "ど"},
	{Set(KeyJ, KeyC), "ば"},
	{Set(KeyJ, KeyX), "び"},
	{Set(KeyF, KeySemi), "ぶ"},
	{Set(KeyF, KeyP), "べ"},
	{Set(KeyJ, KeyZ), "ぼ"},
	{Set(KeyQ), "ゔ"},

	// 半濁音
	{Set(KeyM, KeyC), "ぱ"},
	{Set(KeyM, KeyX), "ぴ"},
	{Set(KeyV, KeySemi), "ぷ"},
	{Set(KeyV, KeyP), "ぺ"},
	{Set(KeyM, KeyZ), "ぽ"},

	// 小書き
	{Set(KeyQ, KeyH), "ゃ"},
	{Set(KeyQ, KeyP), "ゅ"},
	{Set(KeyQ, KeyI), "ょ"},
	{Set(KeyQ, KeyJ), "ぁ"},
	{Set(KeyQ, KeyK), "ぃ"},
	{Set(KeyQ, KeyL), "ぅ"},
	{Set(KeyQ, KeyO), "ぇ"},
	{Set(KeyQ, KeyN), "ぉ"},
	{Set(KeyQ, KeyDot), "ゎ"},
	{Set(KeyG), "っ"},
	{Set(KeyQ, KeyS), "ヶ"},
	{Set(KeyQ, KeyF), "ヵ"},

	// 清音拗音 濁音拗音 半濁拗音
	{Set(KeyR, KeyH), "しゃ"},
	{Set(KeyR, KeyP), "しゅ"},
	{Set(KeyR, KeyI), "しょ"},
	{Set(KeyJ, KeyR, KeyH), "じゃ"},
	{Set(KeyJ, KeyR, KeyP), "じゅ"},
	{Set(KeyJ, KeyR, KeyI), "じょ"},
	{Set(KeyW, KeyH), "きゃ"},
	{Set(KeyW, KeyP), "きゅ"},
	{Set(KeyW, KeyI), "きょ"},
	{Set(KeyJ, KeyW, KeyH), "ぎゃ"},
	{Set(KeyJ, KeyW, KeyP), "ぎゅ"},
	{Set(KeyJ, KeyW, KeyI), "ぎょ"},
	{Set(KeyG, KeyH), "ちゃ"},
	{Set(KeyG, KeyP), "ちゅ"},
	{Set(KeyG, KeyI), "ちょ"},
	{Set(KeyJ, KeyG, KeyH), "ぢゃ"},
	{Set(KeyJ, KeyG, KeyP), "ぢゅ"},
	{Set(KeyJ, KeyG, KeyI), "ぢょ"},
	{Set(KeyD, KeyH), "にゃ"},
	{Set(KeyD, KeyP), "にゅ"},
	{Set(KeyD, KeyI), "にょ"},
	{Set(KeyX, KeyH), "ひゃ"},
	{Set(KeyX, KeyP), "ひゅ"},
	{Set(KeyX, KeyI), "ひょ"},
	{Set(KeyJ, KeyX, KeyH), "びゃ"},
	{Set(KeyJ, KeyX, KeyP), "びゅ"},
	{Set(KeyJ, KeyX, KeyI), "びょ"},
	{Set(KeyM, KeyX, KeyH), "ぴゃ"},
	{Set(KeyM, KeyX, KeyP), "ぴゅ"},
	{Set(KeyM, KeyX, KeyI), "ぴょ"},
	{Set(KeyB, KeyH), "みゃ"},
	{Set(KeyB, KeyP), "みゅ"},
	{Set(KeyB, KeyI), "みょ"},
	{Set(KeyE, KeyH), "りゃ"},
	{Set(KeyE, KeyP), "りゅ"},
	{Set(KeyE, KeyI), "りょ"},

	// 清音外来音 濁音外来音
	{Set(KeyM, KeyE, KeyK), "てぃ"},
	{Set(KeyM, KeyE, KeyP), "てゅ"},
	{Set(KeyJ, KeyE, KeyK), "でぃ"},
	{Set(KeyJ, KeyE, KeyP), "でゅ"},
	{Set(KeyM, KeyD, KeyL), "とぅ"},
	{Set(KeyJ, KeyD, KeyL), "どぅ"},
	{Set(KeyM, KeyR, KeyO), "しぇ"},
	{Set(KeyM, KeyG, KeyO), "ちぇ"},
	{Set(KeyJ, KeyR, KeyO), "じぇ"},
	{Set(KeyJ, KeyG, KeyO), "ぢぇ"},
	{Set(KeyV, KeySemi, KeyJ), "ふぁ"},
	{Set(KeyV, KeySemi, KeyK), "ふぃ"},
	{Set(KeyV, KeySemi, KeyO), "ふぇ"},
	{Set(KeyV, KeySemi, KeyN), "ふぉ"},
	{Set(KeyV, KeySemi, KeyP), "ふゅ"},
	{Set(KeyV, KeyK, KeyO), "いぇ"},
	{Set(KeyV, KeyL, KeyK), "うぃ"},
	{Set(KeyV, KeyL, KeyO), "うぇ"},
	{Set(KeyV, KeyL, KeyN), "うぉ"},
	{Set(KeyM, KeyQ, KeyJ), "ゔぁ"},
	{Set(KeyM, KeyQ, KeyK), "ゔぃ"},
	{Set(KeyM, KeyQ, KeyO), "ゔぇ"},
	{Set(KeyM, KeyQ, KeyN), "ゔぉ"},
	{Set(KeyM, KeyQ, KeyP), "ゔゅ"},
	{Set(KeyV, KeyH, KeyJ), "くぁ"},
	{Set(KeyV, KeyH, KeyK), "くぃ"},
	{Set(KeyV, KeyH, KeyO), "くぇ"},
	{Set(KeyV, KeyH, KeyN), "くぉ"},
	{Set(KeyV, KeyH, KeyDot), "くゎ"},
	{Set(KeyF, KeyH, KeyJ), "ぐぁ"},
	{Set(KeyF, KeyH, KeyK), "ぐぃ"},
	{Set(KeyF, KeyH, KeyO), "ぐぇ"},
	{Set(KeyF, KeyH, KeyN), "ぐぉ"},
	{Set(KeyF, KeyH, KeyDot), "ぐゎ"},
	{Set(KeyV, KeyL, KeyJ), "つぁ"},

	// 記号と制御
	{Set(KeySpace), " "},
	{Set(KeySpace, KeyV), "、"},
	{Set(KeySpace, KeyM), "。"},
	{Set(KeyU), TextBackspace},
	{Set(KeyV, KeyM), TextEnter},
}

// ChordFor は text を入力する組み合わせを返す。
// 別名がある場合は最初に定義されたもの（正規の打鍵）を返す。
func ChordFor(text string) (KeySet, bool) {
	for _, e := range Table {
		if e.Text == text {
			return e.Keys, true
		}
	}
	return 0, false
}

File: web/src/lib/table.ts
// 薙刀式v15のかな変換テーブル。Go版 internal/naginata/table.go の移植。
// 元は eswai/qmk_firmware users/naginata_v15/naginata_v15.c の ngmap。

import { set, type KeySet } from "./keys";

export interface Entry {
  keys: KeySet;
  text: string;
}

export const TEXT_BACKSPACE = "\b";
export const TEXT_ENTER = "\n";

// キー添字 (keys.ts の KEYS 順)
const Q = 0, W = 1, E = 2, R = 3;
const U = 6, I = 7, O = 8, P = 9;
const A = 10, S = 11, D = 12, F = 13, G = 14, H = 15, J = 16, K = 17, L = 18, SEMI = 19;
const Z = 20, X = 21, C = 22, V = 23, B = 24, N = 25, M = 26, COMMA = 27, DOT = 28, SLASH = 29;
const SPACE = 30;

export const TABLE: Entry[] = [
  // 清音
  { keys: set(J), text: "あ" },
  { keys: set(K), text: "い" },
  { keys: set(L), text: "う" },
  { keys: set(SPACE, O), text: "え" },
  { keys: set(SPACE, N), text: "お" },
  { keys: set(F), text: "か" },
  { keys: set(W), text: "き" },
  { keys: set(H), text: "く" },
  { keys: set(S), text: "け" },
  { keys: set(V), text: "こ" },
  { keys: set(SPACE, U), text: "さ" },
  { keys: set(R), text: "し" },
  { keys: set(O), text: "す" },
  { keys: set(SPACE, A), text: "せ" },
  { keys: set(B), text: "そ" },
  { keys: set(N), text: "た" },
  { keys: set(SPACE, G), text: "ち" },
  { keys: set(SPACE, L), text: "つ" },
  { keys: set(E), text: "て" },
  { keys: set(D), text: "と" },
  { keys: set(M), text: "な" },
  { keys: set(SPACE, D), text: "に" },
  { keys: set(SPACE, W), text: "ぬ" },
  { keys: set(SPACE, R), text: "ね" },
  { keys: set(SPACE, J), text: "の" },
  { keys: set(C), text: "は" },
  { keys: set(X), text: "ひ" },
  { keys: set(SPACE, X), text: "ひ" },
  { keys: set(SPACE, SEMI), text: "ふ" },
  { keys: set(P), text: "へ" },
  { keys: set(Z), text: "ほ" },
  { keys: set(SPACE, Z), text: "ほ" },
  { keys: set(SPACE, F), text: "ま" },
  { keys: set(SPACE, B), text: "み" },
  { keys: set(SPACE, COMMA), text: "む" },
  { keys: set(SPACE, S), text: "め" },
  { keys: set(SPACE, K), text: "も" },
  { keys: set(SPACE, H), text: "や" },
  { keys: set(SPACE, P), text: "ゆ" },
  { keys: set(SPACE, I), text: "よ" },
  { keys: set(DOT), text: "ら" },
  { keys: set(SPACE, E), text: "り" },
  { keys: set(I), text: "る" },
  { keys: set(SLASH), text: "れ" },
  { keys: set(SPACE, SLASH), text: "れ" },
  { keys: set(A), text: "ろ" },
  { keys: set(SPACE, DOT), text: "わ" },
  { keys: set(SPACE, C), text: "を" },
  { keys: set(COMMA), text: "ん" },
  { keys: set(SEMI), text: "ー" },

  // 濁音
  { keys: set(J, F), text: "が" },
  { keys: set(J, W), text: "ぎ" },
  { keys: set(F, H), text: "ぐ" },
  { keys: set(J, S), text: "げ" },
  { keys: set(J, V), text: "ご" },
  { keys: set(F, U), text: "ざ" },
  { keys: set(J, R), text: "じ" },
  { keys: set(F, O), text: "ず" },
  { keys: set(J, A), text: "ぜ" },
  { keys: set(J, B), text: "ぞ" },
  { keys: set(F, N), text: "だ" },
  { keys: set(J, G), text: "ぢ" },
  { keys: set(F, L), text: "づ" },
  { keys: set(J, E), text: "で" },
  { keys: set(J, D), text: "ど" },
  { keys: set(J, C), text: "ば" },
  { keys: set(J, X), text: "び" },
  { keys: set(F, SEMI), text: "ぶ" },
  { keys: set(F, P), text: "べ" },
  { keys: set(J, Z), text: "ぼ" },
  { keys: set(Q), text: "ゔ" },

  // 半濁音
  { keys: set(M, C), text: "ぱ" },
  { keys: set(M, X), text: "ぴ" },
  { keys: set(V, SEMI), text: "ぷ" },
  { keys: set(V, P), text: "ぺ" },
  { keys: set(M, Z), text: "ぽ" },

  // 小書き
  { keys: set(Q, H), text: "ゃ" },
  { keys: set(Q, P), text: "ゅ" },
  { keys: set(Q, I), text: "ょ" },
  { keys: set(Q, J), text: "ぁ" },
  { keys: set(Q, K), text: "ぃ" },
  { keys: set(Q, L), text: "ぅ" },
  { keys: set(Q, O), text: "ぇ" },
  { keys: set(Q, N), text: "ぉ" },
  { keys: set(Q, DOT), text: "ゎ" },
  { keys: set(G), text: "っ" },
  { keys: set(Q, S), text: "ヶ" },
  { keys: set(Q, F), text: "ヵ" },

  // 清音拗音 濁音拗音 半濁拗音
  { keys: set(R, H), text: "しゃ" },
  { keys: set(R, P), text: "しゅ" },
  { keys: set(R, I), text: "しょ" },
  { keys: set(J, R, H), text: "じゃ" },
  { keys: set(J, R, P), text: "じゅ" },
  { keys: set(J, R, I), text: "じょ" },
  { keys: set(W, H), text: "きゃ" },
  { keys: set(W, P), text: "きゅ" },
  { keys: set(W, I), text: "きょ" },
  { keys: set(J, W, H), text: "ぎゃ" },
  { keys: set(J, W, P), text: "ぎゅ" },
  { keys: set(J, W, I), text: "ぎょ" },
  { keys: set(G, H), text: "ちゃ" },
  { keys: set(G, P), text: "ちゅ" },
  { keys: set(G, I), text: "ちょ" },
  { keys: set(J, G, H), text: "ぢゃ" },
  { keys: set(J, G, P), text: "ぢゅ" },
  { keys: set(J, G, I), text: "ぢょ" },
  { keys: set(D, H), text: "にゃ" },
  { keys: set(D, P), text: "にゅ" },
  { keys: set(D, I), text: "にょ" },
  { keys: set(X, H), text: "ひゃ" },
  { keys: set(X, P), text: "ひゅ" },
  { keys: set(X, I), text: "ひょ" },
  { keys: set(J, X, H), text: "びゃ" },
  { keys: set(J, X, P), text: "びゅ" },
  { keys: set(J, X, I), text: "びょ" },
  { keys: set(M, X, H), text: "ぴゃ" },
  { keys: set(M, X, P), text: "ぴゅ" },
  { keys: set(M, X, I), text: "ぴょ" },
  { keys: set(B, H), text: "みゃ" },
  { keys: set(B, P), text: "みゅ" },
  { keys: set(B, I), text: "みょ" },
  { keys: set(E, H), text: "りゃ" },
  { keys: set(E, P), text: "りゅ" },
  { keys: set(E, I), text: "りょ" },

  // 清音外来音 濁音外来音
  { keys: set(M, E, K), text: "てぃ" },
  { keys: set(M, E, P), text: "てゅ" },
  { keys: set(J, E, K), text: "でぃ" },
  { keys: set(J, E, P), text: "でゅ" },
  { keys: set(M, D, L), text: "とぅ" },
  { keys: set(J, D, L), text: "どぅ" },
  { keys: set(M, R, O), text: "しぇ" },
  { keys: set(M, G, O), text: "ちぇ" },
  { keys: set(J, R, O), text: "じぇ" },
  { keys: set(J, G, O), text: "ぢぇ" },
  { keys: set(V, SEMI, J), text: "ふぁ" },
  { keys: set(V, SEMI, K), text: "ふぃ" },
  { keys: set(V, SEMI, O), text: "ふぇ" },
  { keys: set(V, SEMI, N), text: "ふぉ" },
  { keys: set(V, SEMI, P), text: "ふゅ" },
  { keys: set(V, K, O), text: "いぇ" },
  { keys: set(V, L, K), text: "うぃ" },
  { keys: set(V, L, O), text: "うぇ" },
  { keys: set(V, L, N), text: "うぉ" },
  { keys: set(M, Q, J), text: "ゔぁ" },
  { keys: set(M, Q, K), text: "ゔぃ" },
  { keys: set(M, Q, O), text: "ゔぇ" },
  { keys: set(M, Q, N), text: "ゔぉ" },
  { keys: set(M, Q, P), text: "ゔゅ" },
  { keys: set(V, H, J), text: "くぁ" },
  { keys: set(V, H, K), text: "くぃ" },
  { keys: set(V, H, O), text: "くぇ" },
  { keys: set(V, H, N), text: "くぉ" },
  { keys: set(V, H, DOT), text: "くゎ" },
  { keys: set(F, H, J), text: "ぐぁ" },
  { keys: set(F, H, K), text: "ぐぃ" },
  { keys: set(F, H, O), text: "ぐぇ" },
  { keys: set(F, H, N), text: "ぐぉ" },
  { keys: set(F, H, DOT), text: "ぐゎ" },
  { keys: set(V, L, J), text: "つぁ" },

  // 記号と制御
  { keys: set(SPACE), text: " " },
  { keys: set(SPACE, V), text: "、" },
  { keys: set(SPACE, M), text: "。" },
  { keys: set(U), text: TEXT_BACKSPACE },
  { keys: set(V, M), text: TEXT_ENTER },
];

// text を入力する組み合わせ。別名は最初に定義されたもの (正規の打鍵)。
const chordIndex = new Map<string, KeySet>();
for (const e of TABLE) {
  if (!chordIndex.has(e.text)) chordIndex.set(e.text, e.keys);
}

export function chordFor(text: string): KeySet | undefined {
  return chordIndex.get(text);
}
